src/ens/records.c

```diff
@@ -0,0 +1,154 @@
+#include <string.h>
+
+#include "../covid_types.h"
+#include "records.h"
+#include "sequencenumber.h"
+#include "storage.h"
+
+int ens_records_iterator_init_range(record_iterator_t* iterator,
+                                    record_sequence_number_t* opt_start,
+                                    record_sequence_number_t* opt_end) {
+    iterator->sn_next = opt_start ? *opt_start : get_oldest_sequence_number();
+    iterator->sn_end = opt_end ? *opt_end : get_latest_sequence_number();
+    if (get_num_records() == 0) {
+        iterator->finished = true;  // no contacts -> no iteration :)
+    } else {
+        iterator->finished = false;
+    }
+    return 0;
+}
+
+/**
+ * Find an entry via binary search for the timestamp.
+ *
+ * @param record pointer to the location, where the loaded record shall be stored
+ * @param target timestamp for which to find the nearest entry for
+ * @param start lower bound for the binary search
+ * @param end upper bound for the binary search
+ */
+int find_record_via_binary_search(record_t* record,
```

Comment:
BTW: Some entries could be deleted or invalid. We also need to handle those cases, i.e. loading the entry "to the left/right".

Reply:
Caution: We need to make sure that this does not deadlock due to missing entries.

Reply:
Check if the oldest ts is newer than the latest -> return error that needs to be handled by the calling function (or the user).